### DISCUSSION TITLE
Avoid drill point for autoheight map for electrolytic plating

### DISCUSSION THREAD

**Discussion by muhammadelmogy** [terjeio/ioSender#237] [Ideas]
The current manufacturing process for PCB engraving using a CNC is fine but it is not really practical if you want to make double sided board.
Imagine a board with 50 vias or even more, you want to solder wire through each via which means more that 100 soldering points and the number doubles for every extra via added to the design.
The current process force you to make the drilling at the end to be able to make the height map which involve touching the surface of the PCB at certain points and to be able to do so you must do this before drilling and that's not accepted when it comes to electrolytic plating for vias!

A more professional process is to drill the PCB first and after that plate the vias using electrolytic plating and after that do the tracing(in our case by engraving).
So we need to some how when doing the height map later for engraving to avoid the drilling points made previously!

What I am thinking right now is to offer the user an option to add the Excellon file for drilling and parse it to know the drills size and positions to make the necessary shift for the height map points.

After that the result will be a grid of point which has been altered to avoid any drill points later!

**Comment by terjeio** [comment]
ioSender can import Excellon files (.drl) and convert them to gcode so it is a matter of adding code to the [height map classes](https://github.com/terjeio/ioSender/tree/master/CNC%20Controls%20Probing/CNC%20Controls%20Probing).
Is there anybody out there who is willing to take on such a challenge?

**Reply by muhammadelmogy** [reply]
> ioSender can import Excellon files (.drl)

Is this happening from the GUI, because I could not find any place to import .drl files!

**Reply by muhammadelmogy** [reply]
I have found this class `Excellon2GCode` under `CNC.Converters` if you mean by import this?

**Reply by muhammadelmogy** [reply]
Ok, after walking through the code (with debugging), I found that you check for the file type using the file extension and for the drill files you are expecting .drl!
If you found .drl , you assume it is an Excellon format file and start parsing and conversion!
The problem in such an approach several thing actually

1. Not all pcb CAD software generate Excellon file with .drl extension! for example, Eagle generate drilling files have xln extension and not drl!
2. Unless the user know about the first point he will never be able to use the Excellon file with ioSender(which happened with me) and he will have to change the file extension to be able to open it!
3. You may or may not agree with me on this one but from design point of view, it is better to make a separate import functionality in the GUI to do this job because the load should only open gcode file and nothing else and taking other format and try to convert it to the gcode format is doing something else that is not very obvious to the user and mix functionalities.
4. From code point of view, from my own experience with programming over the years, I think such an approach make it a little bit harder to maintain the code over years.

That being said, I will proceed with the current situation and try to extract the necessary information from the Excellon file.

**Reply by terjeio** [reply]
1. This can be solved by allowing binding more than one filetype to an importer (by the importer itself).
2. See above.
3. ioSender supports drag/drop of files into the _Program_ and _3D View_ tabs, this is one reasons for how it is implemented.
4. That depends on the code architecture.

**Reply by muhammadelmogy** [reply]
Ok, I have found an issue with parsing drill files generated by Eagle, the parsing couldn't parse drill coordinates and the generated gcode does not have any x,y drill points, after some debugging, I have found that you detect if drilling by checking for G05 which does not exist in the generated Eagle drill file.
File in the attachments
I really do not know whether or not if Eagle is following Excellon file format standards or this is normal and accepted! so should I open an issue with this?

[drill_1_16.zip](https://github.com/terjeio/ioSender/files/9137933/drill_1_16.zip)

**Reply by muhammadelmogy** [reply]
Oh, BTW, I am using FlatCam and it can open the generated drill files from Eagle so I presume the file format for Eagle does not have a problem!

**Reply by muhammadelmogy** [reply]
I added G05 just before first T1 and loaded the file and it came like this in the viewer!
I suppose those moves after each drill action back to the origin is not normal, right?
![ioSenderDrillError](https://user-images.githubusercontent.com/8148383/179680901-aaf4a166-6e52-431c-92ac-1306bb7508d8.jpg)

**Reply by terjeio** [reply]
>  I have found that you detect if drilling by checking for G05 which does not exist in the generated Eagle drill file.

I have only tested with KiCad files - and there are (too?) many ways to produce Exellon files - I'll add support for this format.

> I added G05 just before first T1 and loaded the file and it came like this in the viewer!

I did too and scaling is off. You can fix that for now by [setting the factor to 1000](https://github.com/terjeio/ioSender/blob/29f206a11ba632bbbda509c6aa70a51616a031f1/CNC%20Converters/Excellon2GCode.cs#L180).

> I suppose those moves after each drill action back to the origin is not normal, right?

It is normal, the spindle is moved to the tool change position at 0,0 for each tool.

**Comment by muhammadelmogy** [comment]
I was able to compile the source even the instructions regarding the camera project and removing ADD_CAMERA did not work for me and I ended up adding the AForge lib.
I will try to take a tour in the code and may be with some help from you I can do it(which I highly doubt it since I have no experiences at all with GCode and cnc programming )
